src/engine/init.c

```diff
@@ -246,56 +245,61 @@ modules_load(void)
 	return rc;
 }
 
+static unsigned int
+ncores_needed(unsigned int tgt_nr, unsigned int nr_helpers)
+{
+	return DAOS_TGT0_OFFSET + tgt_nr + nr_helpers;
+}
+
 /**
- * Get the appropriate number of main XS based on the number of cores and
- * passed in preferred number of threads.
+ * Check if the #targets and #nr_xs_helpers is valid to start server, the #nr_xs_helpers possibly
+ * be reduced.
  */
 static int
-dss_tgt_nr_get(unsigned int ncores, unsigned int nr, bool oversubscribe)
+dss_tgt_nr_check(unsigned int ncores, unsigned int tgt_nr, bool oversubscribe)
 {
-	int tgt_nr;
-
 	D_ASSERT(ncores >= 1);
 
 	/* at most 2 helper XS per target */
-	if (dss_tgt_offload_xs_nr > 2 * nr)
-		dss_tgt_offload_xs_nr = 2 * nr;
-	else if (dss_tgt_offload_xs_nr == 0)
+	if (dss_tgt_offload_xs_nr > 2 * tgt_nr) {
+		D_PRINT("#nr_xs_helpers(%d) cannot exceed 2 times #targets (2 x %d = %d).\n",
+			dss_tgt_offload_xs_nr, tgt_nr, 2 * tgt_nr);
+		dss_tgt_offload_xs_nr = 2 * tgt_nr;
+	} else if (dss_tgt_offload_xs_nr == 0) {
 		D_WARN("Suggest to config at least 1 helper XS per DAOS engine\n");
+	}
 
-	/* Each system XS uses one core, and  with dss_tgt_offload_xs_nr
-	 * offload XS. Calculate the tgt_nr as the number of main XS based
-	 * on number of cores.
-	 */
-retry:
-	tgt_nr = ncores - DAOS_TGT0_OFFSET - dss_tgt_offload_xs_nr;
-	if (tgt_nr <= 0)
-		tgt_nr = 1;
-
-	/* If user requires less target threads then set it as dss_tgt_nr,
-	 * if user oversubscribes, then:
-	 *      . if oversubscribe is enabled, use the required number
-	 *      . if oversubscribe is disabled(default),
-	 *        use the number calculated above
-	 * Note: oversubscribing  may hurt performance.
-	 */
-	if (nr >= 1 && ((nr < tgt_nr) || oversubscribe)) {
-		tgt_nr = nr;
-		if (dss_tgt_offload_xs_nr > 2 * tgt_nr)
-			dss_tgt_offload_xs_nr = 2 * tgt_nr;
-	} else if (dss_tgt_offload_xs_nr > 2 * tgt_nr) {
-		dss_tgt_offload_xs_nr--;
-		goto retry;
+	if (oversubscribe) {
+		if (ncores_needed(tgt_nr, dss_tgt_offload_xs_nr) > ncores) {
+			if (ncores > DAOS_TGT0_OFFSET + tgt_nr)
+				dss_tgt_offload_xs_nr = ncores - DAOS_TGT0_OFFSET - tgt_nr;
+			else
+				dss_tgt_offload_xs_nr = 0;
+
+			D_PRINT("Force to start engine with %d targets on %d cores, #nr_xs_helpers "
+				"set as %d.\n",
+				tgt_nr, ncores, dss_tgt_offload_xs_nr);
+		}
+		goto out;
 	}
 
-	if (tgt_nr != nr)
-		D_PRINT("%d target XS(xstream) requested (#cores %d); "
-			"use (%d) target XS\n", nr, ncores, tgt_nr);
+	if (ncores_needed(tgt_nr, dss_tgt_offload_xs_nr) > ncores) {
+		if (ncores < DAOS_TGT0_OFFSET + tgt_nr) {
+			D_ERROR("cannot start engine with %d targets on %d cores, may try with "
+				"DAOS_TARGET_OVERSUBSCRIBE=1\n",
+				tgt_nr, ncores);
+			return -DER_INVAL;
+		}
+		dss_tgt_offload_xs_nr = ncores - DAOS_TGT0_OFFSET - tgt_nr;
+		D_PRINT("Start engine with %d targets on %d cores, #nr_xs_helpers set as %d.\n",
```

Comment:
The message should also indicate that the number of helpers have been forced.

I was expecting that updating this value could only be done with oversubscribe is true.

Reply:
helper threads is not must needed to start daos engine, and start more #helpers than #cores could hurt performance, so I select to reduce #helpers if #cores is not enough but #cores are enough to start the #targets.
@NiuYawei any thought with this?